### ISSUE TITLE
[BUG] Glitches on asrc capture stress test.

### ISSUE THREAD

**Issue by jdrzaszcz** [thesofproject/sof#2981]
**Describe the bug**
When running simultaneous capture and playback on external loopback (same SSP) with ASRC component on capture pipeline, with playback iteration >= 75, there are small gliches on capturead wave. Therefore, the THD value in the wave verification is worse than expected.

**Example captered stream**

Spectogram:
![image](https://user-images.githubusercontent.com/60582003/82986269-d9512880-9ff5-11ea-8f21-b72932b13455.png)

Wave:
![image](https://user-images.githubusercontent.com/60582003/82986407-0dc4e480-9ff6-11ea-9ec5-2ee5fade8be1.png)

**Topology**
```
                +------------------+
    +------+    | +---+  +-------+ |
    | Host |----->|Buf|->|SSP Dai|------------------+
    +------+    | +---+  +-------+ |                |
                +------------------+                |
                                                    |
                +---------------------------------+ |
    +------+    | +---+  +----+  +---+  +-------+ | |
    | Host |<-----|Buf|<-|Asrc|<-|Buf|<-|SSP Dai|<--+
    +------+    | +---+  +----+  +---+  +-------+ |
                +---------------------------------+
```

**To Reproduce**
Ru slim driver python tests from file:
test_01_asrc_simple_capture.py
e.g. 
TestAsrcCapture16000Hz16b16b2ch
TestAsrcCapture48000Hz32b32b2ch
TestAsrcCapture48000Hz16b16b2ch

**Reproduction Rate**
30 %

**Expected behavior**
Captured wave shouldn't have glitches.

**Environment**
I check it on: WHL RVP and TGL RVP,
but as I know from CI tests, it also happens on APL and ICL

Commit: 90ddae9ebe6abbb331a1833a9eacb2d0505865f8 on SOF master branch

**Logs for WHL RVP**
[dma_trace_1.log](https://github.com/thesofproject/sof/files/4715084/dma_trace_1.log)

[19_01_TestAsrcCapture16000Hz16b16b2ch.zip](https://github.com/thesofproject/sof/files/4715092/19_01_TestAsrcCapture16000Hz16b16b2ch.zip)

**Comment by singalsu** [comment]
Interesting, thanks for reporting @jdrzaszcz ! When doing 75+ tests, how long were each play/capture tests?

**Comment by jdrzaszcz** [comment]
@singalsu On each iteration we play 3 seconds stream.

**Comment by singalsu** [comment]
I'm now doing runs of 100x SPDIF ASRC capture test that includes THD+N measurement. I'm seeing variation of about 1 dB but so far no discontinuities. I'll continue tomorrow with more tests.

**Comment by singalsu** [comment]
I caught the first discontinuity, in my test the 3s clip THD+N dropped to -56 dB from usual -78 dB. The waveform looks like this, different looking than what you got. But it's definitely an error if it's not originated from SPDIF transmitter source.

![Screenshot from 2020-06-05 13-05-10](https://user-images.githubusercontent.com/28754793/83864583-9ac21900-a72d-11ea-838a-ef0aa8be223a.png)

**Comment by singalsu** [comment]
I've now run more than 1k of repeats and above has happened only once. Since it looks different than reported and with some kind of impulse train convolved I suspect it was a glitch from SPDIF source (macbook).

I will switch to diagnostic driver test where this reported issue happens. There was earlier also an issue that never happened with Linux driver.

**Comment by singalsu** [comment]
With a lot of help from @jdrzaszcz I have now replicated the issue with my setup and I have an initial fix done & under test. The pull mode filter core has an overrun condition protection that can happen but the SOF buffer update is not aware of the exception and thinks requested amount was processed.

**Comment by singalsu** [comment]
Update - with my test setup I seem to be able to fix the capture glitch issue but there's also for playback a stress test glitch seen for 8 kHz sample rate. It happens only with skew tracking enabled and I can made it disappear almost totally by making the loop filter unpractically slow. The jump in consumed samples originates from HW timestamping register. The samples count delta jumps while time delta remains stable.

So, I have not been able to properly fix it yet. I'll update now the PR with the work for capture, maybe it could be merged before my vacation. If yes, I could create a new issue for playback. As waveform it looks different and more severe with a rather big discontinuity.

**Comment by singalsu** [comment]
There was a regression in PR #3090. While addressing the consume/produce deviation from skew there was a mistake in capture pull mode copy limit for source buffer side. It caused the copy() to process less samples than required. The new PR is #3275.

**Comment by slawblauciak** [comment]
@jdrzaszcz the aforementioned PR has been already merged, can you retest and confirm if it's effective?

**Comment by jdrzaszcz** [comment]
@slawblauciak I will retest it with current master and let you know.

**Comment by jdrzaszcz** [comment]
@singalsu @singalsu Hi, I rerun tests with asrc on capture pipeline and I find out only few results lower than 80 THD value (less than 1 %). But on APL I got few times stream stall on playback pipeline.

**Comment by lgirdwood** [comment]
@jdrzaszcz ok so are we saying a few glitches on APL only but other CAVS1.8+ platforms are OK ?

**Comment by jdrzaszcz** [comment]
@lgirdwood Not exactly.  On WHL, TGL and ICL the results on capture pipeline are fine. I found out only one fail on WHL and one on TGL (but the results on TGL were a little less than 80). I think for this platforms everything is quite good.
On APL RVP i didn't saw glitches, but I found out another problem there (quite often i got stream stall on play stream - about 25 %).

**Comment by jdrzaszcz** [comment]
@singalsu @slawblauciak I rerun tests with SOF master, commit: 32d32b59cb48350f8b159fd63f9f1f74741a7a0e. I run 75 playback iteration for 12 formats (3 sample rate conversion in each format except 8 kHz)
Formats:
![image](https://user-images.githubusercontent.com/46779093/91975936-796acc00-ed20-11ea-9799-abad34acc959.png)

WHL RVP - ones value lower than 80 (~65),
TGL RVP - five value lower than 80 (~ 75),
ICL RVP - two lower than 80(~75);
This means that the reproduction rate is really small, less than 1 %.
Result looks similar for both external and internal loopback.

For APL RVP I got only two times THD value lower than 80 (~60), but I still in about 25 % cases i got stream stall on play stream error.

**Comment by singalsu** [comment]
Thanks @jdrzaszcz I will continue debugging this. The -75 dB THD+N value is about the best quality a C cassette player can ever get so this issue is hard to notice subjectively. I wonder how realistic the test setup where this issue happens is. It would be better to have another real I2S transmitter in the setup that our I2S receiver follows. If the jitter of test setup loop send and receive would accumulate it might put ASRC to worse jitter than normally happens. I'm wondering if using the internal loop force or normal loop is more realistic.

**Comment by lgirdwood** [comment]
@singalsu ping - any update ?

**Comment by lgirdwood** [comment]
@singalsu ping - maybe v1.8 ?

**Comment by singalsu** [comment]
@lgirdwood Yep, unfortunately v1.8. This was confirmed to happen by @jdrzaszcz with two separate I2S interfaces without LBM quirk. So it should be a real issue. I think I should create a glitch detector component, tuned to detect tiny glitches in sine wave (now only a patch for ASRC that didn't trigger) and move it around in the pipeline topology to see where it happens.

**Comment by lgirdwood** [comment]
@singalsu any update ?

**Comment by kkarask** [comment]
I run 19_00_TestAsrcPlayback8000Hz32b32b2ch with playback_iterations=100 and there are glitches in wav files. Reproduction rate 5/100.

Logs:
[2021_05_31_13_22_13.zip](https://github.com/thesofproject/sof/files/6570292/2021_05_31_13_22_13.zip)

**Comment by lgirdwood** [comment]
@kkarask I'm going to close this for the moment as it's not seen with the Linux driver (and wont be seen by users). Lets keep it on record.
